db/fetcher/fetcher.go

```diff
@@ -606,16 +621,46 @@ func (df *DocumentFetcher) fetchNext(ctx context.Context) (EncodedDocument, Exec
 			}
 		}
 
+		// Check if we have read access, for document on this collection, with the given identity.
+		if !df.passedPermissionCheck {
+			if !df.acp.HasValue() {
+				// If no acp module, then we have unrestricted access.
+				df.passedPermissionCheck = true
+			} else {
+				hasPermission, err := permission.CheckAccessOfDocOnCollectionWithACP(
+					ctx,
+					df.identity,
+					df.acp.Value(),
+					df.col,
+					acp.ReadPermission,
+					df.kv.Key.DocID,
+				)
+
+				if err != nil {
+					df.passedPermissionCheck = false
+					return nil, ExecInfo{}, err
+				}
+
+				df.passedPermissionCheck = hasPermission
+			}
+		}
+
 		// if we don't pass the filter (ran and pass)
 		// theres no point in collecting other select fields
 		// so we seek to the next doc
-		spansDone, docDone, err := df.nextKey(ctx, !df.passedFilter && df.ranFilter)
+		spansDone, docDone, err := df.nextKey(ctx, !df.passedPermissionCheck || !df.passedFilter && df.ranFilter)
+
 		if err != nil {
 			return nil, ExecInfo{}, err
 		}
 
-		if docDone {
-			df.execInfo.DocsFetched++
+		if !docDone {
+			continue
+		}
+
+		df.execInfo.DocsFetched++
```

Comment:
I'm going to have to spend a bit more going through the adjusted flow you have here regarding when to `continue` vs when to "normally" proceed. But I dont want to it block my ongoing review, so just putting this comment here as a placeholder.

Reply:
The flow of the fetcher should be the same (with the exception of just now checking with acp)

Reply:
The flow isn't quite the same, as you have inverted the `docDone` and `spansDone` exit/continue conditions. Thats not a bad thing per se, but I just wanted to make sure there wasn't unintended consequences.

Reply:
Resolving please re-raise incase you want to continue or have an action item